--- Day 19: Linen Layout ---
Today, The Historians take you up to the hot springs on Gear Island! Very suspiciously, absolutely nothing goes wrong as they begin their careful search of the vast field of helixes.

Could this finally be your chance to visit the onsen next door? Only one way to find out.

After a brief conversation with the reception staff at the onsen front desk, you discover that you don't have the right kind of money to pay the admission fee. However, before you can leave, the staff get your attention. Apparently, they've heard about how you helped at the hot springs, and they're willing to make a deal: if you can simply help them arrange their towels, they'll let you in for free!

Every towel at this onsen is marked with a pattern of colored stripes. There are only a few patterns, but for any particular pattern, the staff can get you as many towels with that pattern as you need. Each stripe can be white (w), blue (u), black (b), red (r), or green (g). So, a towel with the pattern ggr would have a green stripe, a green stripe, and then a red stripe, in that order. (You can't reverse a pattern by flipping a towel upside-down, as that would cause the onsen logo to face the wrong way.)

The Official Onsen Branding Expert has produced a list of designs - each a long sequence of stripe colors - that they would like to be able to display. You can use any towels you want, but all of the towels' stripes must exactly match the desired design. So, to display the design rgrgr, you could use two rg towels and then an r towel, an rgr towel and then a gr towel, or even a single massive rgrgr towel (assuming such towel patterns were actually available).

To start, collect together all of the available towel patterns and the list of desired designs (your puzzle input). For example:

r, wr, b, g, bwu, rb, gb, br

brwrr
bggr
gbbr
rrbgbr
ubwu
bwurrg
brgr
bbrgwb
The first line indicates the available towel patterns; in this example, the onsen has unlimited towels with a single red stripe (r), unlimited towels with a white stripe and then a red stripe (wr), and so on.

After the blank line, the remaining lines each describe a design the onsen would like to be able to display. In this example, the first design (brwrr) indicates that the onsen would like to be able to display a black stripe, a red stripe, a white stripe, and then two red stripes, in that order.

Not all designs will be possible with the available towels. In the above example, the designs are possible or impossible as follows:

brwrr can be made with a br towel, then a wr towel, and then finally an r towel.
bggr can be made with a b towel, two g towels, and then an r towel.
gbbr can be made with a gb towel and then a br towel.
rrbgbr can be made with r, rb, g, and br.
ubwu is impossible.
bwurrg can be made with bwu, r, r, and g.
brgr can be made with br, g, and r.
bbrgwb is impossible.
In this example, 6 of the eight designs are possible with the available towel patterns.

To get into the onsen as soon as possible, consult your list of towel patterns and desired designs carefully. How many designs are possible?

In [1]:
"""Another path search problem, but this time in string-land."""

'Another path search problem, but this time in string-land.'

In [2]:
with open("input19.txt") as file:
    patterns = next(file).rstrip().split(", ")
    next(file)
    designs = [line.rstrip() for line in file]
patterns.sort(key=len, reverse=True)

In [3]:

def pattern_possible(design):        
    queue = [design]
    visited = set()
    visited.add(design)
    while queue:
        d = queue.pop()
        for p in patterns:
            if d.startswith(p):
                rem_d = d.removeprefix(p)
                if rem_d == "":
                    return True
                elif rem_d not in visited:
                    visited.add(rem_d)
                    queue.append(rem_d)
    return False

In [4]:
num_possible = 0
for d in designs:
    if pattern_possible(d):
        num_possible += 1
num_possible

344

Your puzzle answer was 344.

The first half of this puzzle is complete! It provides one gold star: *

--- Part Two ---
The staff don't really like some of the towel arrangements you came up with. To avoid an endless cycle of towel rearrangement, maybe you should just give them every possible option.

Here are all of the different ways the above example's designs can be made:

brwrr can be made in two different ways: b, r, wr, r or br, wr, r.

bggr can only be made with b, g, g, and r.

gbbr can be made 4 different ways:

g, b, b, r
g, b, br
gb, b, r
gb, br
rrbgbr can be made 6 different ways:

r, r, b, g, b, r
r, r, b, g, br
r, r, b, gb, r
r, rb, g, b, r
r, rb, g, br
r, rb, gb, r
bwurrg can only be made with bwu, r, r, and g.

brgr can be made in two different ways: b, r, g, r or br, g, r.

ubwu and bbrgwb are still impossible.

Adding up all of the ways the towels in this example could be arranged into the desired designs yields 16 (2 + 1 + 4 + 6 + 1 + 2).

They'll let you into the onsen as soon as you have the list. What do you get if you add up the number of different ways you could make each design?

In [26]:
def count_ways(design):
    # search by length of pattern
    scores = [0 for _ in range(len(design) + 1)]
    scores[0] = 1
    for idx in range(len(design)):
        d = design[idx:]
        for p in patterns:
            if d.startswith(p):
                new_idx = idx + len(p)
                scores[new_idx] += scores[idx]
    return scores[-1]

In [27]:
num_ways = 0
for d in designs:
    if pattern_possible(d):
        num_ways += count_ways(d)
num_ways

996172272010026

That's the right answer! You are one gold star closer to finding the Chief Historian.

You have completed Day 19! You can [Shareon Bluesky Twitter Mastodon] this victory or [Return to Your Advent Calendar].